rclcpp/include/rclcpp/context.hpp

```diff
@@ -320,6 +320,11 @@ class Context : public std::enable_shared_from_this<Context>
     return sub_context;
   }
 
+  /// Return the global logging configure mutex.
+  RCLCPP_PUBLIC
+  std::shared_ptr<std::mutex>
+  get_rcl_logging_configure_mutex();
```

Comment:
The shared_ptr seems unnecessary

Reply:
This method should be marked const since it shouldn't mutate the Context

Reply:
I choose to hold a `shared_ptr` to the [mutex](https://github.com/ros2/rclcpp/blob/2aaeee72a6efa3475837245869eb4128f41a57fd/rclcpp/include/rclcpp/context.hpp#L342) in https://github.com/ros2/rclcpp/pull/998 because of the following reason:

- We have the [default context]((https://github.com/ros2/rclcpp/blob/2aaeee72a6efa3475837245869eb4128f41a57fd/rclcpp/include/rclcpp/contexts/default_context.hpp#L48)), with static lifetime.
- The [global logging mutex](https://github.com/ros2/rclcpp/blob/2aaeee72a6efa3475837245869eb4128f41a57fd/rclcpp/src/rclcpp/context.cpp#L42) also have static lifetime.
- According to the standard, objects with static lifetime created in function scope in a single translation unit are created on first use, and destructed in the opposite order they well created. There's no ordering ensured for objects in different translation units.

Then imagine the following situation:
- An user has a program that uses two contexts: the default one and another one with static lifetime.
- One of both contexts get initialized first (doesn't matter which one), [get_global_logging_configure_mutex](https://github.com/ros2/rclcpp/blob/2aaeee72a6efa3475837245869eb4128f41a57fd/rclcpp/src/rclcpp/context.cpp#L40) is called and then the mutex is initialized (on first use).
- The other context is created afterwards.
- As destruction ordering is not ensured between different translation units, a possible destruction order is:
  - context1
  - mutex
  - context2

In case `shutdown` for both contexts wasn't explicitly called the program, the `Context` destructor will call it and that [takes the logging mutex](https://github.com/ros2/rclcpp/blob/2aaeee72a6efa3475837245869eb4128f41a57fd/rclcpp/src/rclcpp/context.cpp#L231), which can segfault if the mutex was previously destroyed.

That's the long story of why using the shared pointer ...

Having say that, I should've added a comment in the code. The rationale is not obvious at all.